src/steps/generic.rs

```diff
@@ -258,10 +258,24 @@ pub fn run_elan(ctx: &ExecutionContext) -> Result<()> {
     let elan = require("elan")?;
 
     print_separator("elan");
-    ctx.run_type()
-        .execute(&elan)
-        .args(["self", "update"])
-        .status_checked()?;
+
+    let disabled = "self-update is disabled";
+    let mut exec = ctx.run_type().execute(&elan);
+    let mut success = true;
+    if let Err(e) = exec.arg("self").arg("update").status_checked() {
```

Comment:
> It still prints the error, which may or may not be desired behaviour.

If you use `output_checked()` directly, then the error won't be printed.

Can you just do, I guess this would work

```rs
let result = ctx
        .run_type()
        .execute(&uv_exec)
        .args(["self", "update"])
        .output_checked();

    if result.is_ok() {
        ctx.run_type()
            .execute(&uv_exec)
            .args(["self", "update"])
            .status_checked()
            .ok();
    }
```

Reply:
This would do self-update twice, if the first one succeeds, right?
Is that better than comparing the output of the "self update" call to a saved error message?

Reply:
> This would do self-update twice, if the first one succeeds, right?

Right, thanks for catching this. I was probably dizzy at that time 🤦‍♂️

-----

> Is that better than comparing the output of the "self update" call to a saved error message?

I think we do need to compare the output with the saved message, what about:

1. Do `elan self update`, capture the output
    1. If 1 succeeds, the self-update feature is available and it just successfully performed the update, we `print!()` the captured `stdout` and `eprint!()` the captured `stderr`
    2. If 1 fails, there are possibly 2 cases why this failure happened:
        1. The self-update feature is unavailable (stdout or stderr contains the saved error message), nothing further needs to be done
        2. The self-update feature is available, but the operation just failed, error out
 2. Do `elan update`

```rust
    // If elan is externally managed, then running `elan self update` will give an 
    // error message that contains `disabled_error_msg`, in this case, Topgrade 
    // tries to hide that `elan self update` has just been executed.
    let disabled_error_msg = "self-update is disabled";
    let executor_output = ctx.run_type().execute(&elan).args(["self", "update"]).output()?;
    match executor_output{
        ExecutorOutput::Wet(command_output) => {
            if command_output.status.success() {
                // Flush the captured output
                std::io::stdout().lock().write(&command_output.stdout).unwrap();
                std::io::stderr().lock().write(&command_output.stderr).unwrap();
            } else {
                let stderr_as_str = std::str::from_utf8(&command_output.stderr).unwrap();
                if stderr_as_str.contains(disabled_error_msg) {
                    // `elan` is externally managed, we cannot do the update. Users
                    // won't see any error message because Topgrade captures them
                    // all.
                } else {
                    // `elan` is NOT externally managed, `elan self update` can 
                    // be performed, but the invocation failed, so we report the
                    // error to the user and error out.
                    std::io::stdout().lock().write(&command_output.stdout).unwrap();
                    std::io::stderr().lock().write(&command_output.stderr).unwrap();

                    return Err(StepFailed.into());
                }
            }

        }
        ExecutorOutput::Dry => {/* nothing needed because in a dry run */}
    }

```